doc/compiler/optimizer/ValuePropagation.md

```diff
@@ -0,0 +1,195 @@
+
+### Store Constraints
+
+There is something special to be said about constraints on stores in the context of global value propagation. In value propagation, the absence of a constraint for a particular value number implies the absence of information about that value number. That is, if we don't know anything about a particular value number we don't generate and store a constraint which says "I don't know anything about the range of values this value number can take". This is done for memory efficiency when intersecting and merging constraint values during propagation.
+
+However consider the following simple while loop:
+
+```
+i = 0;
+
+while (i < 10) {
+    print(i);
+    i = i + 1;
+}
+```
+
+While analyzing this program the value propagation optimization will have created a constraint for the value number of `i` to have a constant constraint of `0`. Later when it encounters the load of `i` as an argument to the `print` function call we need to ensure that we don't fold away the load with the constraint on the value number of `i` which is `0` at that point. This is because we haven't yet processed the second store to `i` which is the loop increment.
+
+To accomplish this the value propagation optimization will ask for the defs (stores which can reach this particular load) of `i` and it will loop through all the defs and check whether they have a constraint. If we encounter a def  which doesn't have a constraint it implies we are in a loop and we haven't yet processed and created all the constraints for `i` and as such we cannot replace the load of `i` with `0` at this point in the analysis.
+
+The absence of a constraint on a store actually means something in this context, which is contrary to what we explained above. It is indeed a special case. Once the analysis reaches the second store, it will update the constraint on the value number for `i` and the second pass through the loop will query whether all the defs for the load of `i` in the `print` function call have a constraint, in which case the answer will be true, and since we have seen the second store the value number of `i` will no longer have a constant constraint of `0`, meaning that we cannot constant fold away the load of `i`.
+
+### Global Constraint Lists
+
+The notion of a global constraint is one which doesn't change for the duration of hte analysis. A typical example of this can be explained by the following program:
+
+```
+i = 5
+
+if (...) {
+    ...
+} else {
+    ...
+}
+```
+
+The value number of the right hand side of the statement `i = 5` has a constraint which will cannot be changed. So why bother to propagate such a constraint through the successor edges of the subsequent `if` statement. Rather than doing this we keep track of a global constraint list to which we add such global constraints which are valid throughout the entire analysis. This saves on memory footprint and time to copy the constraints around as the analysis flows through the method.
+
```

Comment:
"will cannot be changed" is a typo

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/5c1aa757ca3e116307e60404b9d3b0743e2ce03f..127bdeabc6860a9f211446b05d6210a084512605).